lib/graphql-federation.module.ts

```diff
@@ -0,0 +1,157 @@
+import { generateString, extend, mergeDefaults } from './utils';
+import { GraphQLFactory } from './graphql.factory';
+
+@Module({
+  providers: [
+    GraphQLFederationFactory,
+    GraphQLFactory,
+    MetadataScanner,
+    ResolversExplorerService,
+    DelegatesExplorerService,
+    ScalarsExplorerService,
+    GraphQLAstExplorer,
+    GraphQLTypesLoader,
+    GraphQLSchemaBuilder,
+  ],
+  exports: [],
+})
+export class GraphQLFederationModule implements OnModuleInit {
+  private apolloServer: ApolloServer;
+
+  constructor(
+    @Optional()
+    private readonly httpAdapterHost: HttpAdapterHost,
+    @Inject(GRAPHQL_MODULE_OPTIONS)
+    private readonly options: GqlModuleOptions,
+    private readonly graphqlFederationFactory: GraphQLFederationFactory,
+    private readonly graphqlTypesLoader: GraphQLTypesLoader,
+    private readonly graphqlFactory: GraphQLFactory,
+  ) {}
+
+  static forRoot(options: GqlModuleOptions = {}): DynamicModule {
+    options = mergeDefaults(options);
+
+    return {
+      module: GraphQLFederationModule,
+      providers: [
+        {
+          provide: GRAPHQL_MODULE_OPTIONS,
+          useValue: options,
+        },
+      ],
+    };
+  }
+
+  static forRootAsync(options: GqlModuleAsyncOptions): DynamicModule {
+    return {
+      module: GraphQLFederationModule,
+      imports: options.imports,
+      providers: [
+        ...this.createAsyncProviders(options),
+        {
+          provide: GRAPHQL_MODULE_ID,
+          useValue: generateString(),
+        },
+      ],
+    };
+  }
+
+  private static createAsyncProviders(
+    options: GqlModuleAsyncOptions,
+  ): Provider[] {
+    if (options.useExisting || options.useFactory) {
+      return [this.createAsyncOptionsProvider(options)];
+    }
+
+    return [
+      this.createAsyncOptionsProvider(options),
+      {
+        provide: options.useClass,
+        useClass: options.useClass,
+      },
+    ];
+  }
+
+  private static createAsyncOptionsProvider(
+    options: GqlModuleAsyncOptions,
+  ): Provider {
+    if (options.useFactory) {
+      return {
+        provide: GRAPHQL_MODULE_OPTIONS,
+        useFactory: options.useFactory,
+        inject: options.inject || [],
+      };
+    }
+
+    return {
+      provide: GRAPHQL_MODULE_OPTIONS,
+      useFactory: (optionsFactory: GqlOptionsFactory) => optionsFactory.createGqlOptions(),
+      inject: [options.useExisting || options.useClass],
+    };
+  }
+
+  async onModuleInit() {
+    if (!this.httpAdapterHost) return;
+    const { httpAdapter } = this.httpAdapterHost;
+
+    if (!httpAdapter) return;
+
+    const { printSchema } = loadPackage('@apollo/federation', 'ApolloFederation');
+
+    const {
+      path,
+      disableHealthCheck,
+      onHealthCheck,
+      cors,
+      bodyParserConfig,
+      typePaths,
+    } = this.options;
+    const app = httpAdapter.getInstance();
+
+    const typeDefs = await this.graphqlTypesLoader.getTypesFromPaths(typePaths);
```

Comment:
@marcus-sa just a little question, why are u here getting types instead of merging them with mergeTypesByPaths like in graphql.module.ts ?